File: backend/firebrat/utils/ids.py
"""Stable id helpers."""
import hashlib
import os
import re

def sanitize_book_id(pdf_path: str) -> str:
    stem = os.path.splitext(os.path.basename(pdf_path))[0]
    stem = re.sub(r"[^a-zA-Z0-9_-]+", "-", stem).strip("-").lower()
    return stem or "book"

def make_figure_id(n: int) -> str:
    return f"fig_{n:04d}"

def make_formula_id(n: int) -> str:
    return f"formula_{n:04d}"

def make_table_id(n: int) -> str:
    return f"tbl_{n:04d}"

def make_section_id(n: int) -> str:
    return f"sec_{n:04d}"

def make_segment_id(section_id: str, idx: int) -> str:
    # 1-based human ids
    return f"{section_id}_seg_{idx+1:03d}"

def short_hash(text: str, length: int = 8) -> str:
    return hashlib.sha256(text.encode()).hexdigest()[:length]

File: mobile-Backend/lib/src/util/ids.dart
/// Id-minting conventions, mirroring `backend/firebrat/utils/ids.py` exactly
/// (see `docs/DATA_SCHEMA.md`) — the on-device pipeline must produce ids in
/// the same shape as the server pipeline so the two are interchangeable to
/// the reader app, which only ever validates id *format*, not origin.
library;

String sanitizeBookId(String pdfFileName) {
  final stem = pdfFileName.replaceAll(RegExp(r'\.pdf$', caseSensitive: false), '');
  final cleaned = stem
      .replaceAll(RegExp(r'[^a-zA-Z0-9_-]+'), '-')
      .replaceAll(RegExp(r'^-+|-+$'), '')
      .toLowerCase();
  return cleaned.isEmpty ? 'book' : cleaned;
}

String makeFigureId(int n) => 'fig_${n.toString().padLeft(4, '0')}';
String makeFormulaId(int n) => 'formula_${n.toString().padLeft(4, '0')}';
String makeTableId(int n) => 'tbl_${n.toString().padLeft(4, '0')}';
String makeSectionId(int n) => 'sec_${n.toString().padLeft(4, '0')}';

/// 1-based within its section, matching `make_segment_id` server-side.
String makeSegmentId(String sectionId, int zeroBasedIndex) =>
    '${sectionId}_seg_${(zeroBasedIndex + 1).toString().padLeft(3, '0')}';

/// A ref must match one of these shapes to be trusted — anything else is a
/// hallucinated id and gets nulled out rather than propagated, mirroring
/// `compile_llm.py`'s `_sanitize_refs()`.
final RegExp validRefPattern = RegExp(r'^(fig|formula|tbl)_\d{4}$');
